tools/src/tensil/tools/Compiler.scala

```diff
@@ -262,266 +269,428 @@ object Compiler {
           s"No frontend to support ${modelSourceType}"
         )
 
-    val mm = new MemoryManager(
-      constsStream = constsStream,
-      dataType = options.arch.dataType,
-      arch = options.arch,
-      mkConstsDimensions = frontend.mkConstsDimensions,
-      traceContext = traceContext,
-      tracepointConditions = options.tracepointConditions
+    val tempSpace =
+      ArenaMemorySpace("Temp", MemoryTag.Temp, Long.MaxValue)
+    val tempAllocator = new MemoryObjectAllocator(
+      new MemorySpanAllocator()
+    )
+
+    val dram0Space =
+      HeapMemorySpace("DRAM0", MemoryTag.DRAM0, options.arch.dram0Depth)
+    val dram1Space =
+      HeapMemorySpace("DRAM1", MemoryTag.DRAM1, options.arch.dram1Depth)
+    val localSpace =
+      HeapMemorySpace("Local", MemoryTag.Local, options.arch.threadLocalDepth)
+    val freeableAllocator = new MemoryObjectAllocator(
+      new MemorySpanAllocator()
     )
 
     val layout =
       InstructionLayout(options.arch)
+
+    var layerSchedulerResults = mutable.ArrayBuffer.empty[SchedulerResult]
+    var macs                  = 0L
+    var macEfficiency         = 0f
+    val backendStats          = new Stats()
+
+    if (options.printProgress)
+      println(
+        s"Traversing from output node(s): ${outputNames.mkString(",")} ..."
+      )
+
+    val flowNodeNames = frontend.traverse(outputNames)
+
+    if (options.printProgress) {
+      println(s"Found ${flowNodeNames.size} node(s)")
+      println(s"Rewriting emitters ...")
+    }
+
+    val flowEmitters = frontend.rewrite(flowNodeNames)
+
+    if (options.printProgress) {
+      println(s"Rewritten to ${flowEmitters.size} emitter(s)")
+    }
+
+    var nextLayerIndex = 0
+    val schedulingContext = options.strategy match {
+      case CompilerStrategy.LocalIsolated =>
+        new IsolatedLocalSchedulingContext(options)
+      case CompilerStrategy.LocalConsts | CompilerStrategy.LocalVars |
+          CompilerStrategy.LocalVarsAndConsts =>
+        new SharedLocalSchedulingContext(options, localSpace)
+    }
+
+    val mmPass1 = options.strategy match {
```

Comment:
this all wraps up super nicely at the top level. nice job!

Reply:
:sunglasses: